Add DatasetGenerator apply and tests

RyanJHeld opened the pull request:
I figured out the issues I was having, so 1D, 2D, and 3D dataset generation works. I only made value generators for boolean, int, double, and string. It's easy enough to add more, and maybe I'll do that in this PR before merging.

I'm looking for feedback on code organization. Should this new code be separate from the DatasetGenerator code already there? Should the parsing stuff be separate? If so, where should I put it all?

Commit:
Add DatasetGenerator apply and tests

RyanJHeld commented on core/src/test/scala/latis/util/DatasetGenerator.scala:
```diff
@@ -107,3 +196,124 @@ object DatasetGenerator {
+
+  /** Only parses tuples of scalars */
+  def tuple: Parser[modelAst.Model] = for {
+    l <- parens(sepBy(scalar.token, char(',').token))
```
I'm not sure why, but if I try to parse `model`, or even `(tuple | scalar)` here, I get an infinite loop, so this can only parse tuples of scalars.

Commit:
Fix space and newline

lindholc commented:
Regarding code organization: I'd say either rename the file to `datasetGenerator.scala` or make a `generator` subpackage that contains these classes.

lindholc commented on core/src/test/scala/latis/util/DatasetGenerator.scala:
```diff
@@ -107,3 +196,124 @@ object DatasetGenerator {
   }
 }
+
+class BooleanGenerator extends Iterator[Datum] {
```
I wonder if some of these classes could be replaced with something like `Iterator.iterate`.

RyanJHeld replied:
TIL about `Iterator.iterate`. I think I'll use it.

dlindhol commented on core/src/test/scala/latis/util/DatasetGenerator.scala:
```diff
@@ -107,3 +196,124 @@ object DatasetGenerator {
+    t <- valueType
+  } yield modelAst.Scalar(t)
+
+  def valueType: Parser[ValueType] =
```
Could we use `ValueType.fromName` here?

dlindhol commented on core/src/test/scala/latis/util/DatasetGenerator.scala:
```diff
@@ -107,3 +196,124 @@ object DatasetGenerator {
+  } yield StringValueType
+}
+
+object modelAst {
```
Can we simply use `DataType`s here instead of defining a parallel AST? This seems to be a common theme for me: not appreciating an extra layer of abstraction. (Akin to the Operation AST.)

dlindhol commented on core/src/test/scala/latis/util/DatasetGenerator.scala:
```diff
@@ -2,31 +2,120 @@ package latis.util
+   * string. The values, starting with the first sample, increment depending on
+   * their type. Domain variables increment independently.
+   * {{{
+   *   > DatasetGenerator("double -> (double, boolean)")
```
I had imagined a user specifying the model with identifiers: `a -> b` with a default type (Int?)
with the option to specify the types: `a: Int -> b: Double`. That's the direction I was going with `latis.model.ModelParser`.

dlindhol commented on core/src/test/scala/latis/util/DatasetGenerator.scala:
```diff
@@ -2,31 +2,120 @@ package latis.util
+   * ("b", 2) -> (10.0, 11.0)
+   *
+   */
+  def apply(typeSig: String): MemoizedDataset = {
```
Maybe an alternate constructor that takes a `DataType`?

dlindhol commented on core/src/test/scala/latis/util/DatasetGenerator.scala:
```diff
@@ -107,3 +196,124 @@ object DatasetGenerator {
+    case t => throw LatisException(s"No generator for $t")
+  }
+  private val rangeTypes = model.getScalars.drop(model.arity).map(_.valueType)
+  def next(): RangeData = RangeData(rangeTypes.map(nextOfType))
```
If you have two variables of the same type, are they pulling off the same generator/iterator?

Commit:
Move parser stuff to latis.model.ModelParser

and change the parsers to return DataType instead of AST

RyanJHeld commented:
I just committed a refactor.
I took Chris' suggestion about how the iterators are defined. I integrated the parser stuff with Doug's, and kept whatever parser had more functionality. I did a similar thing with the model parser tests. 
I imagine this will have merge conflicts with Shawn's identifier branch.

sapols commented:
> I imagine this will have merge conflicts with Shawn's identifier branch.

I might be wrong, but I actually don’t think I’ve touched any of these files in my branch yet. My changes would definitely produce _errors_ in this code, but I don’t think any conflicts, and the errors would be simple. I’m happy to merge this before my stuff then I could rebase and fix the errors.

RyanJHeld commented:
> > I imagine this will have merge conflicts with Shawn's identifier branch.
> 
> I might be wrong, but I actually don’t think I’ve touched any of these files in my branch yet. My changes would definitely produce _errors_ in this code, but I don’t think any conflicts, and the errors would be simple. I’m happy to merge this before my stuff then I could rebase and fix the errors.

My thought was that the model parsers import `parsers.identifier`, and use it to return a string. Maybe it wouldn't be a merge conflict, but rather a build fail after merging both.

dlindhol commented on core/src/main/scala/latis/model/ModelParser.scala:
```diff
@@ -5,49 +5,62 @@ import Atto._
+  private def valueType: Parser[ValueType] = for {
+    tName <- many(letter)
+    tString = tName.mkString("")
+  } yield ValueType.fromName(tString).fold(throw _, identity)
```
Can we raise an invalid value type into a parse failure?

lindholc replied:
This will throw an exception:

```
@ ModelParser("a: fake -> b: fake") 
latis.util.LatisException: Invalid Scalar value type: fake
```

I'd prefer that this fail in `Parser` rather than throw an exception.

dlindhol commented on core/src/test/scala/latis/model/ModelParserSpec.scala:
```diff
@@ -1,46 +1,110 @@
-     ModelParser.parse("foo") match {
-      case Right(s: Scalar) => s.id should be ("foo")
-    }
+    testScalar("b", Scalar(Metadata("b") + ("type" -> "int")))
```
It's cases like this where I'm inclined to make a convenience constructor for `Scalars`. e.g. 
`apply(id: Identifier, vtype: ValueType)`.
However, that's a slippery slope of combinatorics as we find other properties to include. Then there is the use case. I doubt anyone would be constructing `Scalars` from scratch outside this context of tests. Maybe the DSL? I suspect this will change as we adopt a richer `Metadata` representation.

In other words, I'm looking for opportunities to provide convenience methods but resisting the temptation to pollute the API for the wrong reasons.

dlindhol commented on core/src/test/scala/latis/model/ModelParserSpec.scala:
```diff
@@ -1,46 +1,110 @@
-    ModelParser.parse("(x, y) -> w -> (a, b)") match {
-      case Right(f) => f.toString should be ("(x, y) -> w -> (a, b)")
-    }
+    testFunction("a: int -> b: double -> c: double",
```
How about a nested function with tuples?

dlindhol replied:
Does this support functions in tuples? e.g. `(a, b -> c)`
We don't generally consider this case, but we probably should, some day. Or should we?

dlindhol replied:
Does this support named tuples and functions?
We definitely want names tuples: e.g. `B: (x, y, z)`.
And presumably named functions: e.g. `spectrum: (w -> f)`. (With or without parens?)

dlindhol commented on core/src/test/scala/latis/util/datasetGenerator.scala:
```diff
@@ -197,51 +184,30 @@ object DatasetGenerator {
-    case StringValueType => new StringGenerator
-    case t => throw LatisException(s"No generator for $t")
-  }
+  def stringGenerator: Iterator[Datum] =
```
We should use hurricane names then resort to the greek alphabet. 😜

dlindhol commented on core/src/test/scala/latis/util/datasetGenerator.scala:
```diff
@@ -197,51 +184,30 @@ object DatasetGenerator {
-  }
-}
+  def intGenerator: Iterator[Datum] =
+    Iterator.iterate(IntValue(0))(n => IntValue(n.asInt + 1))
```
I guess it is safe to assume that we won't be hitting max int in this context.

lindholc commented on core/src/test/scala/latis/model/ModelParserSpec.scala:
```diff
@@ -49,4 +113,19 @@ class ModelParserSpec extends FlatSpec {
     }
   }
+
+  implicit def modelEq: DataTypeEquality = new DataTypeEquality {}
```
I compromised when we added the base dataset testing class because it was dragging on, but I don't want to let this pattern perpetuate further.

Normally, one would do something like this:

```scala
implicit val modelEq: Equality[DataType] = new Equality[DataType] {
  // ...
}
```

It's not wrong as it is, but it's unusual.

It would typically live with `DataType`, but because this is for testing it could be defined here (as it already is) or in some package object for type class instances used for tests only.

RyanJHeld replied:
I'll change it and leave it here, and then maybe make a package object when I do the testing refactor.

lindholc reviewed (changes requested):
- I forgot if we decided anything about the parser throwing an exception, but I'd prefer that it didn't.
- We've started an unfortunate precedent for defining type class instances that I think we should stop.

Commit:
Changed valueType parser to not throw an exception

Changed DataTypeEquality type class instance to an implicit value
Ran scala formatter
